core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java

```diff
@@ -0,0 +1,928 @@
+    }
+
+    private boolean checkLowerBound() {
+      RelOptCost upperBound = group.upperBound;
+      if (upperBound.isInfinite()) {
+        return true;
+      }
+      RelOptCost lb = planner.getLowerBound(match);
+      if (upperBound.isLe(lb)) {
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug(
+              "Skip because of lower bound. LB = {}, UP = {}",
+              lb, upperBound);
+        }
+        return false;
+      }
+      return true;
+    }
+
+    @Override public RelSubset group() {
+      return group;
+    }
+
+    @Override public boolean exploring() {
+      return exploring;
+    }
+  }
+
+  private Task getOptimizeInputTask(RelNode rel, RelSubset group) {
+    if (!rel.getTraitSet().satisfies(group.getTraitSet())) {
+      RelNode passThroughRel = convert(rel, group);
+      if (passThroughRel == null) {
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug("Skip optimizing because of traits: {}", rel);
+        }
+        return null;
+      }
+      final RelNode finalPassThroughRel = passThroughRel;
+      applyGenerator(null, () ->
+          planner.register(finalPassThroughRel, group));
+      rel = passThroughRel;
+    }
+    boolean unProcess = false;
+    for (RelNode input : rel.getInputs()) {
+      RelOptCost winner = ((RelSubset) input).getWinnerCost();
+      if (winner == null) {
+        unProcess = true;
+        break;
+      }
+    }
+    if (!unProcess) {
+      return new DeriveTrait(rel, group);
+    }
+    if (rel.getInputs().size() == 1) {
+      return new OptimizeInput1(rel, group);
+    }
+    return new OptimizeInputs(rel, group);
+  }
+
+  private RelNode convert(RelNode rel, RelSubset group) {
+    if (!passThroughCache.contains(rel)) {
+      RelNode passThrough = group.passThrough(rel);
+      if (passThrough != null) {
+        passThroughCache.add(passThrough);
+        return passThrough;
+      }
+    }
+    VolcanoRuleMatch match = ruleQueue.popMatch(
+        Pair.of(rel,
+            m -> m.getRule() instanceof ConverterRule
+                && m.getRule().getOutTrait().satisfies(group.getTraitSet().getConvention())));
+    if (match != null) {
+      tasks.add(new ApplyRule(match, group, false));
+    }
+    return null;
+  }
+
+  /**
+   * O_INPUT when there is only one input
+   */
+  private class OptimizeInput1 implements Task {
+
+    private final RelNode mExpr;
+    private final RelSubset group;
+
+    OptimizeInput1(RelNode mExpr, RelSubset group) {
+      this.mExpr = mExpr;
+      this.group = group;
+    }
+
+
+    @Override public void describe(TaskDescriptor desc) {
+      desc.item("mExpr", mExpr).item("upperBound", group.upperBound);
+    }
+
+    @Override public void perform() {
+      RelOptCost upperBound = group.upperBound;
+      RelOptCost upperForInput = planner.upperBoundForInputs(mExpr, upperBound);
+      if (upperForInput.isLe(planner.zeroCost)) {
+        if (LOGGER.isDebugEnabled()) {
```

Comment:
All the check `if (LOGGER.isDebugEnabled()) {` in this class are redundant.

Reply:
resolved